File: Backend/routes/apiRoutes.js
const express = require('express');
const router = express.Router();
const { Field, Pest, Pump } = require('../models/Schemas');

// GET /api/fields/latest - Retrieve the most recent field data
router.get('/fields/latest', async (req, res) => {
  try {
    const latestField = await Field.findOne().sort({ createdAt: -1 });
    
    if (!latestField) {
      return res.status(404).json({
        success: false,
        message: 'No field data found',
        data: null
      });
    }
    
    res.status(200).json({
      success: true,
      data: latestField
    });
  } catch (error) {
    console.error('Error fetching latest field data:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching field data',
      error: error.message
    });
  }
});

// GET /api/pests - Retrieve pest detection data
router.get('/pests', async (req, res) => {
  try {
    const latestField = await Field.findOne().sort({ createdAt: -1 });
    
    if (!latestField) {
      return res.status(404).json({
        success: false,
        message: 'No field data found. Please add field information first.',
        data: []
      });
    }
    
    const pests = await Pest.find({ fieldId: latestField._id }).sort({ detectedDate: -1 });
    
    if (pests.length === 0) {
      const samplePests = [
        {
          pestName: 'Bollworm',
          severity: 'High',
          affectedArea: '25%',
          detectedDate: new Date('2024-07-10'),
          description: 'Cotton bollworm infestation detected in northern section',
          recommendedTreatment: 'Apply neem-based organic pesticide',
          fieldId: latestField._id
        },
        {
          pestName: 'Aphids',
          severity: 'Medium',
          affectedArea: '15%',
          detectedDate: new Date('2024-07-12'),
          description: 'Aphid colonies found on younger leaves',
          recommendedTreatment: 'Introduce ladybugs as natural predators',
          fieldId: latestField._id
        },
        {
          pestName: 'Whitefly',
          severity: 'Low',
          affectedArea: '8%',
          detectedDate: new Date('2024-07-08'),
          description: 'Minor whitefly presence detected',
          recommendedTreatment: 'Apply yellow sticky traps and neem oil',
          fieldId: latestField._id
        }
      ];
      
      const createdPests = await Pest.insertMany(samplePests);
      return res.status(200).json({
        success: true,
        data: createdPests
      });
    }
    
    res.status(200).json({
      success: true,
      data: pests
    });
  } catch (error) {
    console.error('Error fetching pest data:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching pest data',
      error: error.message
    });
  }
});

// GET /api/pumps - Retrieve pump status data
router.get('/pumps', async (req, res) => {
  try {
    const latestField = await Field.findOne().sort({ createdAt: -1 });
    
    if (!latestField) {
      return res.status(404).json({
        success: false,
        message: 'No field data found. Please add field information first.',
        data: []
      });
    }
    
    const pumps = await Pump.find({ fieldId: latestField._id });
    
    if (pumps.length === 0) {
      const pumpCount = latestField.pesticidePumps || 3;
      const samplePumps = [];
      
      const pumpNames = ['North Section Pump', 'South Section Pump', 'East Section Pump', 'West Section Pump'];
      const pumpStatuses = ['Active', 'Inactive', 'Maintenance'];
      
      for (let i = 0; i < pumpCount; i++) {
        samplePumps.push({
          pumpName: pumpNames[i] || `Pump ${i + 1}`,
          status: pumpStatuses[i % pumpStatuses.length],
          capacity: `${Math.floor(Math.random() * 50) + 50}%`,
          lastMaintenance: new Date(Date.now() - Math.floor(Math.random() * 30) * 24 * 60 * 60 * 1000),
          location: `Section ${String.fromCharCode(65 + i)}`,
          nextMaintenance: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000),
          fieldId: latestField._id
        });
      }
      
      const createdPumps = await Pump.insertMany(samplePumps);
      return res.status(200).json({
        success: true,
        data: createdPumps
      });
    }
    
    res.status(200).json({
      success: true,
      data: pumps
    });
  } catch (error) {
    console.error('Error fetching pump data:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching pump data',
      error: error.message
    });
  }
});

// POST /api/fields - Save field data
router.post('/fields', async (req, res) => {
  try {
    const fieldData = req.body;
    
    // Validate required fields
    const requiredFields = ['cropType', 'area', 'location', 'expectedHarvest', 'pesticidePumps', 'plantingDate'];
    const missingFields = requiredFields.filter(field => !fieldData[field]);
    
    if (missingFields.length > 0) {
      return res.status(400).json({
        success: false,
        message: `Missing required fields: ${missingFields.join(', ')}`
      });
    }
    
    const newField = new Field(fieldData);
    await newField.save();
    
    res.status(201).json({
      success: true,
      message: 'Field data saved successfully',
      data: newField
    });
  } catch (error) {
    console.error('Error saving field data:', error);
    res.status(500).json({
      success: false,
      message: 'Error saving field data',
      error: error.message
    });
  }
});

// GET /api/fields - Get all fields
router.get('/fields', async (req, res) => {
  try {
    const fields = await Field.find().sort({ createdAt: -1 });
    res.status(200).json({
      success: true,
      data: fields
    });
  } catch (error) {
    console.error('Error fetching fields:', error);
    res.status(500).json({
      success: false,
      message: 'Error fetching fields',
      error: error.message
    });
  }
});

// POST /api/pests - Create new pest detection
router.post('/pests', async (req, res) => {
  try {
    const pestData = req.body;
    
    const latestField = await Field.findOne().sort({ createdAt: -1 });
    if (!latestField) {
      return res.status(404).json({
        success: false,
        message: 'No field data found'
      });
    }
    
    pestData.fieldId = latestField._id;
    const newPest = new Pest(pestData);
    await newPest.save();
    
    res.status(201).json({
      success: true,
      message: 'Pest data saved successfully',
      data: newPest
    });
  } catch (error) {
    console.error('Error saving pest data:', error);
    res.status(500).json({
      success: false,
      message: 'Error saving pest data',
      error: error.message
    });
  }
});

module.exports = router;

File: Backend/routes/api_routes.py
from flask import Blueprint, request, jsonify
from models import mongo, Field, Pest, Pump, SeverityEnum, StatusEnum
from bson import ObjectId
from datetime import datetime

api_bp = Blueprint('api', __name__)

@api_bp.route('/fields/latest', methods=['GET'])
def get_latest_field():
    try:
        latest_field = Field.get_collection().find_one(sort=[('createdAt', -1)])
        
        if not latest_field:
            return jsonify({
                'success': False,
                'message': 'No field data found',
                'data': None
            }), 404
        
        return jsonify({
            'success': True,
            'data': Field.to_dict(latest_field)
        }), 200
        
    except Exception as error:
        print(f'Error fetching latest field data: {error}')
        return jsonify({
            'success': False,
            'message': 'Error fetching field data',
            'error': str(error)
        }), 500

@api_bp.route('/pests', methods=['GET'])
def get_pests():
    try:
        latest_field = Field.get_collection().find_one(sort=[('createdAt', -1)])
        
        if not latest_field:
            return jsonify({
                'success': False,
                'message': 'No field data found. Please add field information first.',
                'data': []
            }), 404
        
        pests = list(Pest.get_collection().find(
            {'fieldId': latest_field['_id']}
        ).sort('detectedDate', -1))
        
        return jsonify({
            'success': True,
            'data': [Pest.to_dict(pest) for pest in pests]
        }), 200
        
    except Exception as error:
        print(f'Error fetching pest data: {error}')
        return jsonify({
            'success': False,
            'message': 'Error fetching pest data',
            'error': str(error)
        }), 500

@api_bp.route('/pumps', methods=['GET'])
def get_pumps():
    try:
        latest_field = Field.get_collection().find_one(sort=[('createdAt', -1)])
        
        if not latest_field:
            return jsonify({
                'success': False,
                'message': 'No field data found. Please add field information first.',
                'data': []
            }), 404
        
        pumps = list(Pump.get_collection().find({'fieldId': latest_field['_id']}))
        
        return jsonify({
            'success': True,
            'data': [Pump.to_dict(pump) for pump in pumps]
        }), 200
        
    except Exception as error:
        print(f'Error fetching pump data: {error}')
        return jsonify({
            'success': False,
            'message': 'Error fetching pump data',
            'error': str(error)
        }), 500

@api_bp.route('/fields', methods=['POST'])
def create_field():
    try:
        field_data = request.get_json()
        
        # Validate required fields
        required_fields = ['cropType', 'area', 'location', 'expectedHarvest', 'pesticidePumps', 'plantingDate']
        missing_fields = [field for field in required_fields if field not in field_data or not field_data[field]]
        
        if missing_fields:
            return jsonify({
                'success': False,
                'message': f'Missing required fields: {", ".join(missing_fields)}'
            }), 400
        
        # Prepare data for MongoDB
        field_doc = {
            'cropType': field_data['cropType'],
            'area': field_data['area'],
            'location': field_data['location'],
            'latitude': field_data.get('latitude'),
            'longitude': field_data.get('longitude'),
            'expectedHarvest': field_data['expectedHarvest'],
            'pesticidePumps': field_data['pesticidePumps'],
            'plantingDate': field_data['plantingDate'],
            'soilType': field_data.get('soilType', 'Not specified'),
            'irrigationType': field_data.get('irrigationType', 'Not specified'),
            'createdAt': datetime.utcnow(),
            'updatedAt': datetime.utcnow()
        }
        
        result = Field.get_collection().insert_one(field_doc)
        field_doc['_id'] = str(result.inserted_id)
        
        return jsonify({
            'success': True,
            'message': 'Field data saved successfully',
            'data': field_doc
        }), 201
        
    except Exception as error:
        print(f'Error saving field data: {error}')
        return jsonify({
            'success': False,
            'message': 'Error saving field data',
            'error': str(error)
        }), 500

@api_bp.route('/fields', methods=['GET'])
def get_fields():
    try:
        fields = list(Field.get_collection().find().sort('createdAt', -1))
        
        return jsonify({
            'success': True,
            'data': [Field.to_dict(field) for field in fields]
        }), 200
        
    except Exception as error:
        print(f'Error fetching fields: {error}')
        return jsonify({
            'success': False,
            'message': 'Error fetching fields',
            'error': str(error)
        }), 500

@api_bp.route('/pests', methods=['POST'])
def create_pest():
    try:
        pest_data = request.get_json()
        
        latest_field = Field.get_collection().find_one(sort=[('createdAt', -1)])
        if not latest_field:
            return jsonify({
                'success': False,
                'message': 'No field data found'
            }), 404
        
        # Prepare data for MongoDB
        pest_doc = {
            'fieldId': latest_field['_id'],
            'pestName': pest_data['pestName'],
            'severity': pest_data['severity'],
            'affectedArea': pest_data['affectedArea'],
            'detectedDate': pest_data.get('detectedDate', datetime.utcnow().isoformat()),
            'description': pest_data.get('description'),
            'imageUrl': pest_data.get('imageUrl'),
            'recommendedTreatment': pest_data.get('recommendedTreatment'),
            'status': pest_data.get('status', 'Active'),
            'createdAt': datetime.utcnow(),
            'updatedAt': datetime.utcnow()
        }
        
        result = Pest.get_collection().insert_one(pest_doc)
        pest_doc['_id'] = str(result.inserted_id)
        pest_doc['fieldId'] = str(pest_doc['fieldId'])
        
        return jsonify({
            'success': True,
            'message': 'Pest data saved successfully',
            'data': pest_doc
        }), 201
        
    except Exception as error:
        print(f'Error saving pest data: {error}')
        return jsonify({
            'success': False,
            'message': 'Error saving pest data',
            'error': str(error)
        }), 500


